crates/edr_solidity/src/contracts_identifier/radix_tree.rs

```diff
@@ -0,0 +1,416 @@
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct RadixNode {
+    content: Vec<u8>,
+    is_present: bool,
+    bytes_matched_before: usize,
+    child_nodes: HashMap<u8, RadixNode>,
+}
+
+impl RadixNode {
+    fn new(content: Vec<u8>, is_present: bool, bytes_matched_before: usize) -> RadixNode {
+        RadixNode {
+            content,
+            is_present,
+            bytes_matched_before,
+            child_nodes: HashMap::new(),
+        }
+    }
+
+    pub fn get_content(&self) -> &Vec<u8> {
+        &self.content
+    }
+
+    fn add_word(&mut self, word: Vec<u8>) {
+        if word.is_empty() {
+            return;
+        }
+
+        let b = word[0];
+
+        // we temporarily remove the next node and then insert it back, possibly mutated and/or in a different position
+        let next_node = self.child_nodes.remove(&b);
+
+        match next_node {
+            None => {
+                let bytes_matched_before = self.bytes_matched_before + self.content.len();
+
+                let node = RadixNode::new(word, true, bytes_matched_before);
+
+                self.child_nodes.insert(b, node);
+            }
+            Some(mut next_node) => {
+                let prefix_length = get_shared_prefix_length(&word, &next_node.content);
+
+                // We know it's at least 1
+                assert!(prefix_length > 0);
```

Comment:
I'm wondering how de we know that it's at least 1? Is it because we successfully we removed it?

Reply:
The way this is implemented, the key of an element in `child_nodes` is the first byte of its value, so if `.get(word[0])` exists, then its value must start with `word[0]` and the shared prefix is at least 1.

I'm pretty sure there is a smarter way to implement this that removes this assumption, but I tried to keep the existing implementation as much as possible to get this done sooner.